### PR TITLE
Remove PrivateChannel from caching and events

### PR DESCRIPTION
**Pull request by GnomedDev**
The entire `PrivateChannel` dashmap was never used, discord hasn't supplied events for it since API v8, and all the methods handling it were just unneeded. This allows for removal of `Channel` from a lot of the API surface, replacing it with the specific `GuildChannel` or specifically using `PrivateChannel`.

This was motivated by the `allow(clippy::large_enum_variant)` on `Channel` making me reconsider if it is strictly needed, which it is for a few endpoints (http get channel can return private channels), but was able to be removed from a lot of endpoints.

### CODE REVIEW

**Comment by mkrasnitski** [comment]
This is a rather large change to review this close to 0.12. Even though it's breaking, would you be alright if it was targeted for 0.13 instead? Or is this something you think should absolutely go into 0.12?

**Comment by GnomedDev** [comment]
Knowing serenity's release times, I want this to be usable by users this decade, not next decade.

**Comment by GnomedDev** [comment]
In a more serious note, it's misleading that there is a private channels cache at all, it hasn't worked since serenity bumped past v8, that's when this change should have been made.

File: src/builder/create_invite.rs [hunk 1]
```diff
@@ -284,20 +281,19 @@ impl<'a> Builder for CreateInvite<'a> {
     async fn execute(
         self,
         cache_http: impl CacheHttp,
-        ctx: Self::Context<'_>,
+        channel: Self::Context<'_>,
```
Comment (mkrasnitski) [inline src/builder/create_invite.rs]:
```suggestion
        ctx: Self::Context<'_>,
```
All other builders still name this param `ctx`.

File: src/builder/edit_channel.rs [hunk 2]
```diff
@@ -315,28 +312,18 @@ impl<'a> Builder for EditChannel<'a> {
     async fn execute(
         self,
         cache_http: impl CacheHttp,
-        ctx: Self::Context<'_>,
+        channel: Self::Context<'_>,
```
Comment (mkrasnitski) [inline src/builder/edit_channel.rs]:
```suggestion
        ctx: Self::Context<'_>,
```
Ditto.

File: src/cache/mod.rs [hunk 3]
```diff
@@ -151,7 +149,7 @@ pub struct Cache {
     ///
     /// The TTL for each value is configured in CacheSettings.
     #[cfg(feature = "temp_cache")]
-    pub(crate) temp_channels: MokaCache<ChannelId, GuildChannel, BuildHasher>,
+    pub(crate) temp_channels: MokaCache<ChannelId, Arc<GuildChannel>, BuildHasher>,
```
Comment (mkrasnitski) [inline src/cache/mod.rs:152]: Why was this changed?

**Comment by mkrasnitski** [comment]
What did you use as reference when figuring out when only one of `PrivateChannel` or `GuildChannel` is ever used?

**Comment by GnomedDev** [comment]
Can you ask for a specific method? Most of them are quite obvious, as Webhooks cannot be created in private channels and private channels cannot be cached.

File: src/model/channel/channel_id.rs [hunk 4]
```diff
@@ -415,7 +407,7 @@ impl ChannelId {
         {
             if let Some(cache) = cache_http.cache() {
                 if let Channel::Guild(guild_channel) = &channel {
-                    cache.temp_channels.insert(guild_channel.id, guild_channel.clone());
+                    cache.temp_channels.insert(guild_channel.id, Arc::new(guild_channel.clone()));
```
Comment (mkrasnitski) [inline src/model/channel/channel_id.rs:410]: If `temp_channels` was changed to use `Arc`, why do we still need to clone when we insert into the cache? Is it because the original channel gets dropped here?

File: src/model/channel/guild_channel.rs [hunk 5]
```diff
@@ -354,20 +350,16 @@ impl GuildChannel {
...
-                    Some(self.guild_id),
-                    Permissions::MANAGE_CHANNELS,
-                )?;
+                crate::utils::user_has_perms_cache(cache, self.id, Permissions::MANAGE_CHANNELS)?;
             }
         }
 
-        self.id.delete(cache_http.http()).await
+        let channel = self.id.delete(cache_http.http()).await?;
+        channel.guild().ok_or(Error::Model(ModelError::InvalidChannelType))
```
Comment (mkrasnitski) [inline src/model/channel/guild_channel.rs:362]: The error here is technically unreachable (we're a `GuildChannel` here), but meh, probably fine.

File: src/model/channel/private_channel.rs [hunk 6]
```diff
@@ -69,8 +69,8 @@ impl PrivateChannel {
     /// closing a private channel on the client, which can be re-opened.
     #[allow(clippy::missing_errors_doc)]
     #[inline]
-    pub async fn delete(&self, http: impl AsRef<Http>) -> Result<Channel> {
-        self.id.delete(http).await
+    pub async fn delete(&self, http: impl AsRef<Http>) -> Result<PrivateChannel> {
+        self.id.delete(http).await?.private().ok_or(Error::Model(ModelError::InvalidChannelType))
```
Comment (mkrasnitski) [inline src/model/channel/private_channel.rs:73]: Same here as for `GuildChannel`, we can probably assert that this won't ever produce an invalid channel.

File: src/model/webhook.rs [hunk 7]
```diff
@@ -191,28 +194,32 @@ impl WebhookChannel {
...
+                            .insert(guild_channel.id, Arc::new(guild_channel.clone()));
+                    }
                 }
-            }
-        }
 
-        Ok(channel)
+                Ok(guild_channel)
+            },
+            other => Err(Error::Model(ModelError::NonGuildWebhookChannel(Box::new(other)))),
```
Comment (mkrasnitski) [inline src/model/webhook.rs]:
```suggestion
            other => Err(Error::Model(ModelError::InvalidChannelType)),
```
Couldn't we just use `InvalidChannelType` here and call it a day? If it's not possible to make a webhook for a private channel, it's not worth creating an error variant that will never realistically be constructed at runtime.

File: src/cache/event.rs [hunk 8]
```diff
@@ -38,113 +38,52 @@ use crate::model::user::{CurrentUser, OnlineStatus};
...
-            },
-        }
-
-        old_channel
+        cache
+            .guilds
+            .get_mut(&self.channel.guild_id)
+            .and_then(|mut g| g.channels.insert(self.channel.id, self.channel.clone()))
     }
 }
```
Comment (mkrasnitski) [inline src/cache/event.rs:79]: Having trouble reviewing this because the diff is quite messy. But, I'm assuming changing `.map` to `.and_then` works because `HashMap::insert` returns the old value wrapped in `Some` (and `None` if there is no old value)?

It's clever, but maybe it's worth adding a comment to that effect so that other people can better understand what's happening without much effort.

**Comment by mkrasnitski** [comment]
> Can you ask for a specific method? Most of them are quite obvious, as Webhooks cannot be created in private channels and private channels cannot be cached.

My main concern was about webhooks, but that's been answered in converstion in the serenity discord. Almost done with review here.

File: src/cache/mod.rs [hunk 9]
```diff
@@ -469,75 +445,6 @@ impl Cache {
...
-    /// # async fn run() -> Result<(), Box<dyn std::error::Error>> {
-    /// let mut client =
-    ///     Client::builder("token", GatewayIntents::default()).event_handler(Handler).await?;
-    ///
-    /// client.start().await?;
-    /// # Ok(())
-    /// # }
-    /// ```
-    ///
-    /// [`EventHandler::message`]: crate::client::EventHandler::message
```
Comment (mkrasnitski) [inline src/cache/mod.rs:520]: Do we want to keep this example and just move it onto `Cache::channel`?

File: src/cache/mod.rs [hunk 10]
```diff
@@ -387,39 +362,40 @@ impl Cache {
...
-        {
-            if let Some(channel) = self.temp_channels.get(&id) {
-                return Some(Channel::Guild(channel));
-            }
-        }
-
-        if let Some(private_channel) = self.private_channels.get(&id) {
-            return Some(Channel::Private(private_channel.clone()));
+        if let Some(channel) = self.temp_channels.get(&id) {
+            return Some(CacheRef::from_arc(channel));
```
Comment (mkrasnitski) [inline src/cache/mod.rs]: Minor nit, but for diff cleanliness purposes, let's keep the wrapping braces around this. Serenity is inconsistent with how it chooses to feature-gate single-expression blocks.  Sometimes it wraps them in braces and sometimes not. But the behavior is equivalent so it's something to be decided later.

File: src/cache/mod.rs [hunk 11]
```diff
@@ -387,39 +362,40 @@ impl Cache {
...
-    /// This will search the `channels` map, then the [`Self::private_channels`] map.
-    ///
-    /// If you know what type of channel you're looking for, you should instead manually retrieve
-    /// from one of the respective methods:
+    /// Retrieves a [`GuildChannel`] from the cache based on the given Id.
     ///
-    /// - [`GuildChannel`]: [`Self::guild_channel`] or [`Self::guild_channels`]
-    /// - [`PrivateChannel`]: [`Self::private_channel`] or [`Self::private_channels`]
+    /// This will search the guild cache for the [`Guild::channels`] map, then the
+    /// `Self::temp_channels` map if enabled and not found.
```
Comment (mkrasnitski) [inline src/cache/mod.rs]: This should be reworded since `Cache::temp_channels` isn't a method and the field is only `pub(crate)`. Plus, it's feature-gated.

**Review by mkrasnitski — CHANGES_REQUESTED** [review]

File: src/cache/mod.rs (hunk 3 above)
Comment (GnomedDev) [reply src/cache/mod.rs:152]: This allows `Cache::channel` to return a `CacheRef`, as `MokaCache` will clone the value when `get` is called. If not wrapped, `Cache::channel` would either have to clone guild channels in the more popular path or `CacheRef` would have to include owned values, which feels very wrong.

File: src/cache/mod.rs (hunk 9 above)
Comment (GnomedDev) [reply src/cache/mod.rs:520]: It seems largely useless as an example, as `Message` already has a `to_channel` method, but if you feel strongly about this example I'll move it to `Cache::channel`.

File: src/model/channel/channel_id.rs (hunk 4 above)
Comment (GnomedDev) [reply src/model/channel/channel_id.rs:410]: We do need to clone before inserting into the cache. Arc::new takes an owned value and we are returning back the channel we just inserted. A clever thought would be to use `CacheRef` to always refer back to either the `cache.channel` or the `temp_channels` map, but this falls apart when `temp_cache` is disabled.

File: src/model/channel/guild_channel.rs (hunk 5 above)
Comment (GnomedDev) [reply src/model/channel/guild_channel.rs:362]: The alternative would be a panic, and in exceptional circumstances I still prefer an error.

File: src/model/channel/private_channel.rs (hunk 6 above)
Comment (GnomedDev) [reply src/model/channel/private_channel.rs:73]: See above.

File: src/model/channel/channel_id.rs (hunk 4 above)
Comment (mkrasnitski) [reply src/model/channel/channel_id.rs:410]: Right, the lifetime of the channel changes if we enable or disable `temp_cache`.

File: src/cache/event.rs (hunk 8 above)
Comment (GnomedDev) [reply src/cache/event.rs:79]: I kind of agree with you here, but looking at the whole module sees this trick used quite a couple of times, so I don't think it's necessary to comment here.

**Review by mkrasnitski — APPROVED** [review]
LGTM

**Comment by mkrasnitski** [comment]
Last thing - do we also want to rename `Cache::guild_channels` to `Cache::channels`, since we essentially did the same with the singular version `Cache::guild_channel` -> `Cache::channel`?

EDIT: It's probably fine to leave it, since `channel` takes a `ChannelId`, but `guild_channels` takes a `GuildId`, and all the other methods that do so also have a `guild_` prefix.

**Review by arqunis — APPROVED** [review]
